Sources/audio-devices/AudioDevices.swift

```diff
@@ -1,181 +1,302 @@
-  var volume: Double?
-
+  let hasInput: Bool
+  let hasOutput: Bool
+  
   init(withId deviceId: AudioDeviceID) throws {
     id = deviceId
     var deviceName = "" as CFString
     var deviceUID = "" as CFString
-
+    
     do {
       try CoreAudioData.getAudioData(id: deviceId, selector: kAudioObjectPropertyName, value: &deviceName)
     } catch {
       throw AudioDevices.Error.invalidDeviceId
     }
-
+    
     name = deviceName as String
-
+    
     do {
       try CoreAudioData.getAudioData(id: deviceId, selector: kAudioDevicePropertyDeviceUID, value: &deviceUID)
     } catch {
       throw AudioDevices.Error.invalidDeviceId
     }
-
+    
     uid = deviceUID as String
-
+    
     let inputChannels: UInt32 = try CoreAudioData.getAudioDataSize(
       id: deviceId,
       selector: kAudioDevicePropertyStreams,
       scope: kAudioDevicePropertyScopeInput
     )
-
-    isInput = inputChannels > 0
-
     let outputChannels: UInt32 = try CoreAudioData.getAudioDataSize(
       id: deviceId,
       selector: kAudioDevicePropertyStreams,
       scope: kAudioDevicePropertyScopeOutput
     )
-
-    isOutput = outputChannels > 0
-
-    let hasVolume = CoreAudioData.hasAudioData(
-      id: deviceId,
-      selector: kAudioDevicePropertyVolumeScalar,
-      scope: kAudioDevicePropertyScopeOutput
-    )
-
-    if hasVolume {
+    
+    hasInput = inputChannels > 0
+    hasOutput = outputChannels > 0
+  }
+  
+  func getInputVolume() -> Double? {
+    do {
       var deviceVolume: Float32 = 0.0
-
-      do {
-        try CoreAudioData.getAudioData(
-          id: deviceId,
-          selector: kAudioDevicePropertyVolumeScalar,
-          scope: kAudioDevicePropertyScopeOutput,
-          value: &deviceVolume
-        )
-
-        volume = Double(deviceVolume)
-      } catch {
-        volume = nil
-      }
-    } else {
-      volume = nil
+      try CoreAudioData.getAudioData(
+        id: id,
+        selector: kAudioDevicePropertyVolumeScalar,
+        scope: kAudioDevicePropertyScopeInput,
+        value: &deviceVolume
+      )
+      
+      return Double(deviceVolume)
+    } catch {
+      return nil
     }
   }
-
-  mutating func setVolume(_ newVolume: Double) throws {
-    guard volume != nil else {
-      throw AudioDevices.Error.volumeNotSupported
+  
+  func getOutputVolume() -> Double? {
+    do {
+      var deviceVolume: Float32 = 0.0
+      try CoreAudioData.getAudioData(
+        id: id,
+        selector: kAudioHardwareServiceDeviceProperty_VirtualMasterVolume,
+        scope: kAudioDevicePropertyScopeOutput,
+        value: &deviceVolume
+      )
+      
+      return Double(deviceVolume)
+    } catch {
+      return nil
     }
-
+  }
+  
+  func getVolume() -> Double? {
+    // TODO: need to think how to handle when hasInput && hasOutput
+    if hasInput {
+      return getInputVolume()
+    }
+    if hasOutput {
+      return getOutputVolume()
+    }
+    
+    return nil
+  }
+  
+  func getInputMute() -> Bool? {
```

Comment:
```suggestion
  var isInputMuted: Bool? {
```

Same pattern for the others.